### PR TITLE
feat: delete for me logic without UI

### LINKED ISSUE #13873: Implement “Delete for me” option for messages
# Feature Issue

### User Story

As a user, I want to delete messages locally **so that** I can remove message I don't want to see without removing them to the other chat participants. 

### Description

Flow TBD

### Acceptance Criteria

TBD

### Notes

Link to Message Context Drawer in Figma: https://www.figma.com/file/PPWkgOYlZZDxZv5SDGsZVV/Posts-%26-Attachments?node-id=4388%3A631636

### PR DESCRIPTION
**Pull request by yqrashawn**
related #13873

- [x] menu UI (will implement in a seperate PR)
- [ ] deleted message UI (will implement in a seperate PR)
- [ ] deleted for me notification with undo UI (will implement in a seperate PR)
- [x] delete/undo logic
- [x] status-go https://github.com/status-im/status-go/pull/2866 (version 0.111.1 already included in status-mobile)
- [x] temp UI indicating message deleted for me (deleted for me messages have a red border)

### Steps to test
- Open Status A with credential 1
- Open Status B with credential 1
- Open Status C with credential 2
- Pair A and B
- Switch A B C to the new UI
- Send message1 from 2 to 1
- (Press and hold message to get the `Delete for me` button)
- (Deleted for me message will have a red border around it)
- Delete message1 on A
	- message1 should be `deleted for me` on A, B but not C after 4+ seconds, showing
		- delete for me action has a 4 second (not the final number) delay before actually persist into db and synced to paired devices
		- delete for me can be applied to message send by others
		- delete for me only delete for **me**(for credential 1, which is status A and B)
		- delete for me message can be synced between paired devices

status: ready

### CODE REVIEW

**Comment by yqrashawn** [comment]
`DeleteMessageAndSend` https://github.com/status-im/status-go/blob/0170d50702d0e3532d794e104cd0eb0a0687fb14/protocol/messenger_messages.go#L86 has the code to delete message and [handle `chat.LastMessage`](https://github.com/status-im/status-go/blob/0170d50702d0e3532d794e104cd0eb0a0687fb14/protocol/messenger_messages.go#L148)
[HandleDeleteMessage](https://github.com/status-im/status-go/blob/0170d50702d0e3532d794e104cd0eb0a0687fb14/protocol/messenger_handler.go#L1112)

**Comment by yqrashawn** [comment]
Relies on 
1. toast component https://github.com/status-im/status-mobile/issues/13831
2. system message https://github.com/status-im/status-mobile/issues/13939
3. message context drawer https://github.com/status-im/status-mobile/issues/13527

File: src/status_im/ui/screens/chat/message/message.cljs [hunk 1]
```diff
@@ -1,34 +1,35 @@
 (ns status-im.ui.screens.chat.message.message
-  (:require [re-frame.core :as re-frame]
+  (:require [quo.core :as quo]
+            [quo.design-system.colors :as colors]
+            [re-frame.core :as re-frame]
+            [reagent.core :as reagent]
+            [status-im.chat.models.delete-message-for-me]
```
Comment (yqrashawn) [inline src/status_im/ui/screens/chat/message/message.cljs:6]: rest of the ns form is only reorders

File: src/status_im/ui/screens/chat/message/message.cljs [hunk 2]
```diff
@@ -320,12 +322,17 @@
...
-                                   :margin-top 2}
+                                   :margin-top   2}
                                   (style/message-timestamp-text))
             :accessibility-label :message-timestamp}
            timestamp-str]])
-     ;;MESSAGE CONTENT
-       content
+       ;; MESSAGE CONTENT
+       ;; TODO(yqrashawn): wait for system message component to display deleted for me UI
+       (if deleted-for-me?
```
Comment (yqrashawn) [inline src/status_im/ui/screens/chat/message/message.cljs:331]: Temporally add a red border for `deleted for me` messages for QA

File: src/status_im/utils/config.cljs [hunk 3]
```diff
@@ -182,3 +182,5 @@
 ;; usable enough to replace the old one **in the new UI**.
 (def new-activity-center-enabled?
   (atom false))
+
+(def delete-message-for-me-undo-time-limit-ms 4000)
```
Comment (yqrashawn) [inline src/status_im/utils/config.cljs:186]: Not the final number

File: src/status_im/chat/models/delete_message_for_me.cljs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+               :deleted-for-me-undoable-till)))
+
+(defn- update-db-delete-message-for-me-locally
+  [db chat-id message-id undo-time-limit-ms]
+  (when (get-in db [:messages chat-id message-id])
+    (update-in db
+               [:messages chat-id message-id]
+               assoc
+               :deleted-for-me? true
+               :deleted-for-me-undoable-till (+ (js/Date.now)
```
Comment (ilmotta) [inline src/status_im/chat/models/delete_message_for_me.cljs]: We have a utility namespace for date & time operations. I think you can replace this call and the other one in `update-db-undo-delete-message-for-me-locally` with the function `status-im.utils.datetime/timestamp`.

File: src/status_im/chat/models/delete_message_for_me.cljs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  [{:keys [db]} {:keys [chat-id message-id]} undo-time-limit-ms]
+  (when (get-in db [:messages chat-id message-id])
+    {:db (update-db-delete-message-for-me-locally db chat-id message-id undo-time-limit-ms)
+
+     :utils/dispatch-later
+     [{:dispatch [:chat.ui/delete-message-for-me-and-sync chat-id message-id]
+       :ms       undo-time-limit-ms}]}))
+
+(fx/defn undo-delete-message-for-me
+  "Undo delete for me, only allowed in limited time"
```
Comment (ilmotta) [inline src/status_im/chat/models/delete_message_for_me.cljs]: I didn't quite understand this docstring. I know what you mean, but only because I looked into the source of `update-db-undo-delete-message-for-me-locally`. I would suggest rewording it or removing it. I found the implementation to be clear already without the docstring.

File: src/status_im/chat/models/delete_message_for_me.cljs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  {:events [:chat.ui/delete-message-for-me-and-sync]}
+  [{:keys [db]} {:keys [message-id chat-id]}]
+  (when (get-in db [:messages chat-id message-id])
+    {:db
+     (update-db-clear-delete-message-for-me-undo-timer db chat-id message-id)
+     ::json-rpc/call
+     [{:method "wakuext_deleteMessageForMeAndSync"
+       :params [chat-id message-id]
+       :js-response true
+       :on-error #(log/error "failed to delete message message for me " %)
```
Comment (ilmotta) [inline src/status_im/chat/models/delete_message_for_me.cljs]: Just a typo, the word `message` is duplicated.

File: src/status_im/chat/models/delete_message_for_me.cljs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    {:db (update-db-undo-delete-message-for-me-locally db chat-id message-id)}))
+
+(fx/defn delete-message-for-me-and-sync
+  {:events [:chat.ui/delete-message-for-me-and-sync]}
+  [{:keys [db]} {:keys [message-id chat-id]}]
+  (when (get-in db [:messages chat-id message-id])
+    {:db
+     (update-db-clear-delete-message-for-me-undo-timer db chat-id message-id)
+     ::json-rpc/call
+     [{:method "wakuext_deleteMessageForMeAndSync"
```
Comment (ilmotta) [inline src/status_im/chat/models/delete_message_for_me.cljs]: [Code style] You could align the map keys here.

File: src/status_im/chat/models/delete_message_for_me_test.cljs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+   [status-im.chat.models.delete-message-for-me :as to-test]))
+
+(defonce mid "message-id")
+(defonce cid "chat-id")
+
+(deftest delete-for-me
+  (let [db      {:messages {cid {mid {:id mid}}}}
+        message {:message-id mid :chat-id cid}]
+    (testing
+     "delete for me"
```
Comment (ilmotta) [inline src/status_im/chat/models/delete_message_for_me_test.cljs]: [Code style] I think it's more common to leave the test description in the same line as the `testing` macro, both in status-mobile codebase and in the wild.

File: src/status_im/chat/models/delete_message_for_me.cljs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+(defn- update-db-clear-delete-message-for-me-undo-timer
+  [db chat-id message-id]
+  (when (get-in db [:messages chat-id message-id])
+    (update-in db
+               [:messages chat-id message-id]
+               dissoc
+               :deleted-for-me-undoable-till)))
+
+(defn- update-db-delete-message-for-me-locally
```
Comment (ilmotta) [inline src/status_im/chat/models/delete_message_for_me.cljs]: You could take advantage of Clojure's namespacing capabilities and not repeat the words `delete-message-for-me` in all vars, given the namespace already says it's all related to `delete-message-for-me`. For example (just a suggestion for the names):

- `update-db-delete-message-for-me-locally` becomes `update-db-locally`
- `update-db-clear-delete-message-for-me-undo-timer` becomes `update-db-clear-undo-timer`
- `undo-delete-message-for-me` becomes `undo`
- `delete-message-for-me` becomes `delete`
- `delete-message-for-me-and-sync` becomes `delete-and-sync`

**Review by ilmotta — APPROVED** [review]
LGTM @yqrashawn, and I really enjoyed seeing the tests for the critical logic in the event handlers.

File: src/status_im/chat/models/delete_message_for_me.cljs (hunk 5 above)
Comment (yqrashawn) [reply src/status_im/chat/models/delete_message_for_me.cljs]: removed the doc string, add new doc string to `update-db-undo-locally`

File: src/status_im/chat/models/delete_message_for_me.cljs (hunk 4 above)
Comment (yqrashawn) [reply src/status_im/chat/models/delete_message_for_me.cljs]: done

File: src/status_im/chat/models/delete_message_for_me.cljs (hunk 7 above)
Comment (yqrashawn) [reply src/status_im/chat/models/delete_message_for_me.cljs]: Do you mean like this?
```clojure
{:db             (update-db-clear-undo-timer db chat-id message-id)
 ::json-rpc/call [{:method      "wakuext_deleteMessageForMeAndSync"
                   :params      [chat-id message-id]
                   :js-response true
                   :on-error    #(log/error "failed to delete message for me " %)
                   :on-success  #(re-frame/dispatch [:sanitize-messages-and-process-response %])}]}
```

File: src/status_im/chat/models/delete_message_for_me.cljs (hunk 9 above)
Comment (yqrashawn) [reply src/status_im/chat/models/delete_message_for_me.cljs]: done

File: src/status_im/chat/models/delete_message_for_me_test.cljs (hunk 8 above)
Comment (yqrashawn) [reply src/status_im/chat/models/delete_message_for_me_test.cljs]: I agree, this is more friendly to grep. 
I found zprint want to put it into a newline. Let me check if there's anyway to tweak zprint to respect the same line in `testing` form

File: src/status_im/chat/models/delete_message_for_me.cljs (hunk 7 above)
Comment (ilmotta) [reply src/status_im/chat/models/delete_message_for_me.cljs]: That looks probably more in line with what the rest of the team prefers.

File: src/status_im/chat/models/delete_message_for_me_test.cljs (hunk 8 above)
Comment (ilmotta) [reply src/status_im/chat/models/delete_message_for_me_test.cljs]: I see, you're already using zprint. I think it gets confused with the line length. It shouldn't break the line for certain macros like this one, which expect a description that will often go past 80 characters long.

**Comment by flexsurfer** [comment]
@yqrashawn it seems like it was implemented for the old UI ?

**Comment by qoqobolo** [comment]
Hey @yqrashawn, thanks for your work!
Could you clarify please, what is the expected behavior you mentioned in this scenario:
> Steps to test
Open Status A with credential 1
Open Status B with credential 1
Open Status C with credential 2
Pair A and B
Switch A B C to the new UI
Send message1 from 2 to 1
(Press and hold message to get the Delete for me button)
(Deleted for me message will have a red border around it)
Delete message1 on A
message1 should be deleted for me on A, B but not C showing
delete for me can be applied to message send by others
delete for me only delete for me(for credential 1, which is status A and B)
delete for me message can be synced between paired devices

Currently, a `deleted for me` message has a red border around it on the Status A device, but it is not deleted on Status B.
Should the deleted message be synced in fact or it will be implemented next?

**Comment by yqrashawn** [comment]
> @yqrashawn it seems like it was implemented for the old UI ?

@flexsurfer  It works on old UI. It's just there's no delete for me button in the old UI since it's not added into `message_old.cljs`


> Should the deleted message be synced in fact or it will be implemented next?

@qoqobolo  It should be synced between Status A and Status B if they have the same credential and are paired with each other.

**Comment by qoqobolo** [comment]
@yqrashawn okay, there is an issue then, can you take a look, please? 

### ISSUE 1: Message is not deleted on synced device

**STR:**
1. Recover the same user (same credentials) on DeviceA and DeviceB
2. Pair DeviceA and DeviceB
3. Enable New UI on both devices
4. Join a public chat on DeviceA
5. Send a message
6. `Delete for me` the message on DeviceA
7. Check DeviceB

**Expected result:** the message is deleted on DeviceB as well
**Actual result:** the message has a red border on DeviceA and NOT deleted on DeviceB

**Logs:**
DeviceA [geth.log](https://github.com/status-im/status-mobile/files/9754609/geth.log)
DeviceB [geth.log](https://github.com/status-im/status-mobile/files/9754618/geth.log)

https://user-images.githubusercontent.com/67952253/195080849-b045d096-528e-4770-8553-6b45bdd23c94.mp4

**Comment by yqrashawn** [comment]
@qoqobolo Fixed. 


Caused by 
1. Calling `:chat.ui/delete-message-for-me-and-sync` with wrong arguments
2. The status-go version in current dev branch doesn't have the delete for me feature yet.

Also note that there's a 4-second delay before status-mobile send the delete for me action to status-go (there will be a button to undo delete for me in the future and user can only undo the action within a time limit (which is 4 seconds for now)), so there'll be at least 4+ seconds before another paired device got synced.

**Comment by qoqobolo** [comment]
Thanks @yqrashawn, LGTM.
PR is tested and can be merged.
